### ISSUE TITLE
HI3516CV100 IMX291 support

### ISSUE THREAD

**Issue by DaxonBaggus** [OpenIPC/firmware#669]
XiongMai made an IPC module with HI3516CV100 SoC and IMX291 sensor.

Strangely the IMX291 is supported on many other HI3516 chips, but not on this one. Could something be added or changed to allow this combination to work?

XM product ID: IPG-54H20PL-S
PCB Marking: BLK18C-0090-38X38_S
Sensor: IMX291
FM version: V4.02.R11.00002532.10010.242800.00000
SoC: HI3516CV100

I have around 20 cameras I'm trying to refurbish with OpenIPC so maybe I can help with a donation?

![IMG_20230313_222235](https://user-images.githubusercontent.com/4037783/224851032-6160a56d-4be5-44e3-90ec-7fbf3b3a8a89.jpg)
![IMG_20230313_222259](https://user-images.githubusercontent.com/4037783/224851045-86ee1a5c-1b14-4e66-9a7f-392f2601b801.jpg)

**Comment by widgetii** [comment]
IMX291 sensor driver should be written from scratch for this model. It will take 4-6 weeks of full-time reverse-engineer work with close interaction with hardware.

I doubt that it's even reasonable judging you can replace all boards to modern version

**Comment by DaxonBaggus** [comment]
Thanks for your honest answer. Then sadly I will have to look for a different solution until someone feels the need to go and make this happen.

**Comment by FlyRouter** [comment]
I wrote you some questions in PM Telegram, please post as you get a chance.

**Comment by FlyRouter** [comment]
I made a small crutch with imx138 since we don't have a driver.
Not sure if it will work well but it breathes nonetheless... ;)
https://openipc.org/snapshots/368807

**Comment by FlyRouter** [comment]
Solution

One commit for [sensor init](https://github.com/OpenIPC/firmware/commit/b3b507242557b545a67e0cd0134112e99fd1a3f4)

Set old integrated sensor from Linux
`fw_setenv sensor imx138_spi`

**Comment by DaxonBaggus** [comment]
[IPG-54H20PL-S stock firmwares 2532.zip](https://github.com/OpenIPC/firmware/files/11011215/IPG-54H20PL-S.stock.firmwares.2532.zip)

Attached all the versions of the stock firmware for this board. Maybe this contains a driver somewhere? 

The camera briefly worked with the imx138 change, but after changing a couple of settings in the web interface it is now broken, and I have not been able to get it do display an image again.

**Comment by widgetii** [comment]
Sensor driver is compiled into Sofia binary statically. There is no way to extract and use it

**Comment by DaxonBaggus** [comment]
Ah ok, I think I understand now, you are looking for something like this?
https://github.com/EliasKotlyar/Xiaomi-Dafang-Software/blob/master/drivers/sensors/imx291/imx291.c

**Comment by widgetii** [comment]
No, this is driver for Ingenic platform (the code is absolutely useless for HiSilicon)

**Comment by mpissarello** [comment]
Same issue as #204 
At November 2022, my camera was half-working, see [my comment](https://github.com/OpenIPC/firmware/issues/204#issuecomment-1305774560), I need to try with the latest OpenIPC firmware.

**Comment by openipc-ai** [comment]
Closing as a duplicate of #204. As noted in the thread above it is the same board (54H20PL-S, Hi3516CV100 + IMX291) and the same symptom.

Keeping the discussion in one place — #204 has the longer history, so please continue there.
